src/vidpf.rs

```diff
@@ -226,6 +228,67 @@ impl<W: VidpfValue, const NONCE_SIZE: usize> Vidpf<W, NONCE_SIZE> {
         })
     }
 
+    /// [`Vidpf::eval_cached`] evaluates the entire `input` and produces a share of the
+    /// input's weight. It reuses computation from previous levels available in the
+    /// cache
+    pub fn eval_cached(
+        &self,
+        key: &VidpfKey,
+        public: &VidpfPublicShare<W>,
+        input: &VidpfInput,
+        cache_tree: &mut BinaryTree<VidpfEvalCache<W>>,
+        nonce: &[u8; NONCE_SIZE],
+    ) -> Result<VidpfValueShare<W>, VidpfError> {
+        let n = input.len();
+        if n > public.cw.len() {
+            return Err(VidpfError::InvalidAttributeLength);
+        }
+
+        let state = VidpfEvalState::init_from_key(key);
+        let path = input;
+        match cache_tree.get(input.prefix(0).index.as_bitslice()) {
+            Some(_) => (),
+            None => cache_tree
+                .insert(
+                    IdpfInput::empty_input().index.as_bitslice(),
+                    self.eval_next_cached(key.id, public, input, 0, &state, nonce)?,
+                )
+                .expect("inserting into top of empty tree"),
+        };
```

Comment:
Instead of calling `get()` and then `insert()` you should be able to call `insert()` and handle the error if the data already exists.